src/sdk/namespace_fs.js

```diff
@@ -513,7 +534,36 @@ class NamespaceFS {
                 if (r.common_prefix) {
                     res.common_prefixes.push(r.key);
                 } else {
-                    res.objects.push(this._get_object_info(bucket, r.key, r.stat, 'null'));
+                    const obj_info = this._get_object_info(bucket, r.key, r.stat, 'null', true);
+                    if (!list_versions && obj_info.delete_marker) {
+                        continue;
+                    }
+                    if (list_versions) {
+                        const obj_info_list = [];
+                        versions.forEach(v => {
+                            let search_key = v.key.replace(HIDDEN_VERSIONS_PATH, '');
+                            if (search_key.startsWith('/')) {
+                                search_key = search_key.replace('/', '');
+                            } else {
+                                search_key = search_key.replace('//', '/');
+                            }
+                            if (search_key.startsWith(r.key)) {
+                                obj_info_list.push(this._get_object_info(bucket, r.key, v.stat, 'null', false));
+                            }
+                        });
+                        /* arrange the object versions in ascending create time */
+                        obj_info_list.sort(function(o1, o2) {
```

Comment:
do you think we can add it to the dir_cache as we do for the other sort? 
sort by time will be called if the dir is some_path/.versions/ ?

Reply:
If I am not wrong, in `process_dir()`, we enter each dir and process them into a list. 

ex: assume `dir1` `dir2` are 2 dirs, under which we have `.versions` dir. When we enter `dir1`, we process the entries. If we sort the objects under `.versions` then we sort all entries irrespective of versions of particular key. So I thought not to go for sort here and filter the versions outside based on the key, push them to `obj_info_list` and sort them based on time.

Reply:
aha I understand, can we maybe sort first by key and then by create time? 
because here we will always need to sort even though we could have these entries using cache

Reply:
We can skip the sort based on `key` for '.versions' dir.
Whether we sort here or outside number of sorting cycles wont differ. 
Once we have the list ready (cached), we can run the sorting for `.versions` once the `stat` for the objects available.

Reply:
the downside of not sorting it inside the cache is that the cache won't keep the entries sorted so we will need to sort the versions every time the user calls list_object_version,
A question - 
is there any specific reason for sorting by c_time (file's metadata, such as its ownership, location, file type, and permission settings change time) for sorting instead of sorting by m_time (content modification time)? 
I'm wondering if we can maybe instead of stat take the m_time from the version's key, and then you can sort it inside the cache
Tagging @guymguym if you want to share an opinion here

Reply:
Following an online review with @v-harihar and @romayalon:
* in the current implementation, there is a nested loop over all dir entries and all versions under `.versions`. this can have scale issues when there are a lot of files and versions and needs to be changed.
* As Romy suggested above, we will keep `.versions` sorted in the cache. Sort order is first by key and then by version time. 
* once we have both the dir entries and the versions in 2 sorted lists, we can merge them together, without the need for nested loops or resorting
* after merging the 2 lists, we need to take into account the `key_marker`, the `version_id_marker` and `limit` for returning the actual list result